src/pycell.rs

```diff
@@ -0,0 +1,446 @@
+//! Traits and structs for `#[pyclass]`.
+use crate::conversion::{AsPyPointer, FromPyPointer, ToPyObject};
+use crate::pyclass_init::PyClassInitializer;
+use crate::pyclass_slots::{PyClassDict, PyClassWeakRef};
+use crate::type_object::{PyDowncastImpl, PyObjectLayout, PyObjectSizedLayout, PyTypeInfo};
+use crate::types::PyAny;
+use crate::{ffi, FromPy, PyClass, PyErr, PyNativeType, PyObject, PyResult, Python};
+use std::cell::{Cell, UnsafeCell};
+use std::fmt;
+use std::mem::ManuallyDrop;
+use std::ops::{Deref, DerefMut};
+
+#[doc(hidden)]
+#[repr(C)]
+pub struct PyCellBase<T: PyTypeInfo> {
+    ob_base: T::Layout,
+    borrow_flag: Cell<BorrowFlag>,
```

Comment:
Is it worth making an assertion that this doesn't need to be thread-safe (i.e. not `AtomicXyz`) because the of the GIL?

Reply:
This is already not `!Send` nor `Sync`.
What kind of assertion do you have in mind?

Reply:
Even if it's not `Send` or `Sync` in Rust, the python interpreter will happily share this data between threads. Perhaps just a comment explaining why `Cell` is enough?

Reply:
I see.